File: test/VbNet.LanguageServer.Tests/Services/HoverServiceTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Services;
using VbNet.LanguageServer.Workspace;
using Xunit;

namespace VbNet.LanguageServer.Tests.Services;

/// <summary>
/// Unit tests for HoverService.
/// </summary>
public class HoverServiceTests
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly HoverService _hoverService;

    public HoverServiceTests()
    {
        _workspaceManager = new WorkspaceManager(NullLogger<WorkspaceManager>.Instance);
        _documentManager = new DocumentManager(_workspaceManager, NullLogger<DocumentManager>.Instance);
        _hoverService = new HoverService(
            _workspaceManager,
            _documentManager,
            NullLogger<HoverService>.Instance);

        _workspaceManager.Initialize();
    }

    [Fact]
    public async Task GetHoverAsync_NoDocument_ReturnsNull()
    {
        var @params = new HoverParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = "file:///nonexistent.vb" },
            Position = new Position { Line = 0, Character = 0 }
        };

        var result = await _hoverService.GetHoverAsync(@params, CancellationToken.None);

        Assert.Null(result);
    }

    [Fact]
    public async Task GetHoverAsync_StandaloneDocument_ReturnsNull()
    {
        var uri = "file:///c:/test/module.vb";
        var text = "Module Module1\n    Sub Main()\n    End Sub\nEnd Module";

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = uri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var @params = new HoverParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = uri },
            Position = new Position { Line = 1, Character = 8 }
        };

        // Without a workspace, we can't get hover info
        var result = await _hoverService.GetHoverAsync(@params, CancellationToken.None);

        Assert.Null(result);
    }

    [Fact]
    public async Task GetHoverAsync_NullParams_ReturnsNull()
    {
        var result = await _hoverService.GetHoverAsync(null!, CancellationToken.None);

        Assert.Null(result);
    }

    [Fact]
    public async Task GetHoverAsync_NullTextDocument_ReturnsNull()
    {
        var @params = new HoverParams
        {
            TextDocument = null!,
            Position = new Position { Line = 0, Character = 0 }
        };

        var result = await _hoverService.GetHoverAsync(@params, CancellationToken.None);

        Assert.Null(result);
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Services/HoverServiceTests.vb
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Services
Imports VbNet.LanguageServer.Workspace
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Services

    ''' <summary>
    ''' Unit tests for HoverService.
    ''' </summary>
    Public Class HoverServiceTests
        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _hoverService As HoverService

        Public Sub New()
            _workspaceManager = New WorkspaceManager(NullLogger(Of WorkspaceManager).Instance)
            _documentManager = New DocumentManager(_workspaceManager, NullLogger(Of DocumentManager).Instance)
            _hoverService = New HoverService(
                _workspaceManager,
                _documentManager,
                NullLogger(Of HoverService).Instance)

            _workspaceManager.Initialize()
        End Sub

        <Fact>
        Public Async Function GetHoverAsync_NoDocument_ReturnsNull() As Task
            Dim request = New HoverParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = "file:///nonexistent.vb"},
                .Position = New Position With {.Line = 0, .Character = 0}
            }

            Dim result = Await _hoverService.GetHoverAsync(request, CancellationToken.None)

            Assert.Null(result)
        End Function

        <Fact>
        Public Async Function GetHoverAsync_NullParams_ReturnsNull() As Task
            Dim result = Await _hoverService.GetHoverAsync(Nothing, CancellationToken.None)

            Assert.Null(result)
        End Function

        <Fact>
        Public Async Function GetHoverAsync_StandaloneDocument_ReturnsNull() As Task
            Dim uri = "file:///c:/test/module.vb"
            Dim text = "Module Module1" & vbLf & "    Sub Main()" & vbLf & "    End Sub" & vbLf & "End Module"

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = uri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim request = New HoverParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = uri},
                .Position = New Position With {.Line = 1, .Character = 8}
            }

            Dim result = Await _hoverService.GetHoverAsync(request, CancellationToken.None)

            Assert.Null(result)
        End Function

        <Fact>
        Public Async Function GetHoverAsync_NullTextDocument_ReturnsNull() As Task
            Dim request = New HoverParams With {
                .TextDocument = Nothing,
                .Position = New Position With {.Line = 0, .Character = 0}
            }

            Dim result = Await _hoverService.GetHoverAsync(request, CancellationToken.None)

            Assert.Null(result)
        End Function
    End Class

End Namespace
